### PR TITLE
feat(hydroflow_lang): Implement partition operator

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Supports both named ports and numeric indices.

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
TODO: operator doc and more compile-fail tests

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+use super::{
+    FlowProperties, FlowPropertyVal, OperatorCategory, OperatorConstraints, PortListSpec,
+    WriteContextArgs, RANGE_0, RANGE_1,
+};
+use crate::diagnostic::{Diagnostic, Level};
+use crate::graph::ops::OperatorWriteOutput;
+use crate::graph::{OperatorInstance, PortIndexValue};
+use crate::pretty_span::PrettySpan;
+
+/// This oeprator takes the input pipeline and allows the user to determine which singular output
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:18]: typo

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+/// my_partition[rest] -> for_each(|v| println!("{}", v));
+/// ```
+///
+/// # Indexed mode
+/// With indexed mode, the closure's second argument is a single usize value equal to the number
+/// of output pipelines. Each output pipeline port must be numbered with an index, starting from
+/// zero and with no gaps. The closure returns the index of the desired output port.
+///
+/// ```hydroflow
+/// my_partition = source_iter(1..=100) -> partition(|v, n| v % n);
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:52]: where does `n` come from in this snippet? Confusing.

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+    op_name: &'static str,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> Result<Option<Vec<Ident>>, ()> {
+    // Port idents supplied via port connections in the surface syntax.
+    // Two modes, either all numeric `0, 1, 2, 3, ...` or all `Ident`s.
+    // If ports are `Idents` then the closure's 2nd argument, input array must have named
+    // values corresponding to the port idents.
+    let mut ports_numeric = BTreeSet::new();
+    let mut ports_idents = Vec::new();
+    // If any ports are elided we return `Err(())` early.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:194]: "elided" is a fancy word and kind of vague. is an "elided" port an expression like the second line below?
```
demux_out[a] -> ...;
demux_out -> ... ;
```
Maybe say "if a demux operator is used without naming an output port" or something?

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    let mut err_elided = false;
+    for output_port in output_ports {
+        match output_port {
+            PortIndexValue::Elided(port_span) => {
+                err_elided = true;
+                diagnostics.push(Diagnostic::spanned(
+                    port_span.unwrap_or(op_span),
+                    Level::Error,
+                    format!("Output ports from `{}` cannot be elided.", op_name),
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:203]: if you change in the docs also change in the error messages.

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+                op_name
+            ),
+        ));
+    }
+
+    // Port idents specified in the closure's second argument.
+    let mut arg2 = &mut func.inputs[1];
+
+    // Port idents specified in the closure's second argument.
+    // let mut arg2 = &mut func.inputs[1];
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:302]: remove commented-out code

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+/// pipeline each item should be delivered to.
+///
+/// > Arguments: A Rust closure, the first argument is a reference to the item and the second
+/// argument corresponds to one of two modes, either named or indexed.
+///
+/// > Note: The closure has access to the [`context` object](surface_flows.md#the-context-object).
+///
+/// # Named mode
+/// With named ports, the closure's second argument must be an array pattern of names, such as
+/// `[foo, bar, baz]`, where each name is an output port. The closure should return the name of the
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:28]: what is an "array pattern of names"? Is that something specific in Rust?

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    if let Pat::Ident(pat_ident) = arg2 {
+        arg2 = &mut *pat_ident
+            .subpat
+            .as_mut()
+            .ok_or_else(|| Diagnostic::spanned(
+                arg2_span,
+                Level::Error,
+                format!(
+                    "Second argument for the `{}` closure must contain a slice sub-pattern to determine the port names and order. \
+                    For example: `arr @ [foo, bar, baz]` for ports `foo`, `bar`, and `baz`.",
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:319]: `must contain a Rust 'slice sub-pattern'`

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+            ))?
+            .1;
+    }
+    let Pat::Slice(pat_slice) = arg2 else {
+        return Err(Diagnostic::spanned(
+            arg2_span,
+            Level::Error,
+            format!(
+                "Second argument for the `{}` closure must have a slice pattern to determine the port names and order. \
+                For example: `[foo, bar, baz]` for ports `foo`, `bar`, and `baz`.",
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:331]: "have a Rust slice pattern"

File: hydroflow/tests/surface_demux_partition.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        my_partition[fzbz] -> for_each(|_| println!("fizzbuzz"));
+    };
+    df.run_available();
+}
+
+#[multiplatform_test]
+pub fn test_partition_round() {
+    let mut df = hydroflow_syntax! {
+        my_partition = source_iter(0..100)
+            -> partition(|v, len| v % len);
```
Comment (jhellerstein) [inline hydroflow/tests/surface_demux_partition.rs:105]: I'm confused where `len` comes from!

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 10]
```diff
@@ -82,19 +82,19 @@ pub const DEMUX: OperatorConstraints = OperatorConstraints {
...
         if 2 != func.inputs.len() {
             diagnostics.push(Diagnostic::spanned(
-                op_span,
+                func.span(),
                 Level::Error,
                 &*format!(
-                    "Closure provided as second argument to `{}` must have a second \
-                    input argument listing ports, `var_args!(port_a, port_b, ...)`.",
+                    "Closure provided to `{}(..)` must have two arguments: \
+                    the first argument is the item, and the second argument lists ports, e.g. `var_args!(port_a, port_b, ..)`.",
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:97]: "e.g. the second item could be `var_args...`"

File: hydroflow_lang/src/graph/ops/partition.rs [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        );
+    };
+    if 2 != func.inputs.len() {
+        return Err(Diagnostic::spanned(
+            func.span(),
+            Level::Error,
+            &*format!(
+                "Closure provided to `{}(..)` must have two arguments: \
+                the first argument is the item, and for named ports the second argument must contain a slice sub-pattern to determine the port names and order. \
+                For example: `[foo, bar, baz]` for ports `foo`, `bar`, and `baz`.",
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/partition.rs:292]: "For example, the second argument could be `[foo...`"

File: hydroflow/tests/compile-fail/surface_demux_badarg.stderr [hunk 12]
```diff
@@ -0,0 +1,13 @@
+error: Argument must be a two-argument closure expression
+ --> tests/compile-fail/surface_demux_badarg.rs:5:48
+  |
+5 |         my_demux = source_iter(0..10) -> demux(std::mem::drop);
+  |                                                ^^^^^^^^^^^^^^
+
+warning: unused import: `var_args`
```
Comment (jhellerstein) [inline hydroflow/tests/compile-fail/surface_demux_badarg.stderr:7]: This is kind of a distraction

**Review by jhellerstein — APPROVED** [review]
Some suggestions on documentation and error messages inline.

File: .vscode/settings.json [hunk 13]
```diff
@@ -3,6 +3,7 @@
         {
             // Set output levels for `tracing` logging.
             "env": {
+                "RUST_BACKTRACE": "full",
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply .vscode/settings.json:6]: Yup, shows more error info when running tests from VS code

**Pull request merged**